### PR TITLE
Datasource parameter extension

### LINKED ISSUE #296: Modifying datasource endpoints before execution
Many remote APIs require parameters to be included in the endpoint path, rather than in the querystring.

In the current version of Web, we can use requestParams (parameters extracted from the URL) to modify an endpoint, but in some cases it's required to modify an endpoint using data previously fetched.

Rather than extend the requestParams functionality, I think we need something akin to filterEvent to allow modifying the endpoint from an event.

This is somewhat similar to @adamkdean's issue: https://github.com/dadi/web/issues/216

Perhaps both chaining _and_ modifying via an event should be possible.

### PR DESCRIPTION
**Pull request by jimlambie**
As noted in #296, this PR modifies the way request params are obtained, making them available from the request, config or the session (if one exists). Perhaps requestParams isn't the right name for them now, but we could shift the perspective that they are _for the next request_ rather than _from the request_.

- [x] extract parameters from request (default)
- [x] extract parameters from config
- [x] extract parameters from session
- [x] extract parameters from chained datasource
- [x] modify endpoint from event

### Request params from different sources

```
 "requestParams": [
      {
        "source": "config",
        "param": "global.values.token",
        "field": "token",
        "target": "endpoint"
      },
      {
        "source": "session",
        "param": "user.profile._id",
        "field": "user",
        "target": "endpoint"
      },
      {
        "param": "title",
        "field": "title",
        "target": "filter"
      }
    ]
```

### Inject params from previous datasources

In addition, chained datasources can now use data from the datasource they're attached to, modifying the endpoint if `"target": "endpoint"`:

```
{
  "datasource": {
    "key": "chained-endpoint",
    "source": {
      "type": "remote",
      "protocol": "http",
      "host": "127.0.0.1",
      "port": "3000",
      "endpoint": "1.0/makes/{name}"
    },
    "chained": {
    "datasource": "global",   // the ds we're waiting on
    "outputParam": {
      "param": "results.0.name",  // where to find the value in the global ds
      "field": "name",   // the value to replace in this endpoint 
      "target": "endpoint"  // obvious, really
    }
  }
}
```

### Modify datasource endpoint using Events

To modify a datasource's endpoint using an event, add a property to the datasource `"endpointEvent": "filename-of-event.js"` and return the new endpoint in the callback: `callback(null, 'v1/twitter.json')`

## Closure!

Close #296 , #216 , #198

### CODE REVIEW

File: dadi/lib/datasource/index.js [hunk 1]
```diff
@@ -1,20 +1,22 @@
 /**
  * @module Datasource
  */
-var _ = require('underscore')
-var fs = require('fs')
-var path = require('path')
-var url = require('url')
+const _ = require('underscore')
```
Comment (eduardoboucas) [inline dadi/lib/datasource/index.js]: Could we use this as an opportunity to get rid of Underscore on this file? 😬 Only 3 instances I can see.

Comment (jimlambie) [reply dadi/lib/datasource/index.js]: Absolutely!

Comment (abovedave) [reply dadi/lib/datasource/index.js]: 🎉

Comment (abovedave) [reply dadi/lib/datasource/index.js]: #198

Comment (jimlambie) [reply dadi/lib/datasource/index.js]: That issue should have a list of files that still contain underscore usage :)

**Comment by jimlambie** [comment]
@adamkdean please re-review - have added endpoint modification for chained datasources

File: dadi/lib/controller/index.js [hunk 2]
```diff
@@ -140,18 +142,16 @@ Controller.prototype.buildInitialViewData = function (req) {
   if (urlData.pathname.length) data.pathname = urlData.pathname
 
   // add request params (params from the path, e.g. /:make/:model)
-  _.extend(data.params, req.params)
-
   // add query params (params from the querystring, e.g. /reviews?page=2)
-  _.extend(data.params, data.query)
+  data.params = Object.assign({}, req.params, data.query)
```
Comment (adamkdean) [inline dadi/lib/controller/index.js]: Much nicer approach

File: dadi/lib/controller/index.js [hunk 3]
```diff
@@ -178,11 +178,11 @@ Controller.prototype.process = function process (req, res, next) {
...
 
-  var statusCode = res.statusCode || 200
-  var data = this.buildInitialViewData(req)
-  var view = new View(req.url, this.page, data.json)
+  const statusCode = res.statusCode || 200
+  let data = this.buildInitialViewData(req)
+  const view = new View(req.url, this.page, data.json)
 
-  var done = data.json
+  let done = data.json
```
Comment (adamkdean) [inline dadi/lib/controller/index.js]: Is `done` mutated at all? If not, shouldn't it be a const?

File: dadi/lib/controller/index.js [hunk 4]
```diff
@@ -296,34 +296,27 @@ Controller.prototype.loadEventData = function (events, req, res, data, done) {
 }
 
 Controller.prototype.loadData = function (req, res, data, done) {
-  var self = this
+  const self = this
 
-  var primaryDatasources = {}
-  var chainedDatasources = {}
+  let primaryDatasources = {}
+  let chainedDatasources = {}
```
Comment (adamkdean) [inline dadi/lib/controller/index.js]: Do these two variables ever get reassigned away from the `{}` object they begin with? If it's only the object they reference that mutates, we can still use `const` here.

Comment (jimlambie) [reply dadi/lib/controller/index.js]: true

File: test/app/datasources/car-makes-chained-endpoint.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
+{
+	"datasource": {
+		"key": "car-makes-chained-endpoint",
+		"name": "Makes datasource",
+		"source": {
+			"type": "remote",
+			"protocol": "http",
+			"host": "127.0.0.1",
+			"port": "3000",
+			"endpoint": "1.0/makes/{name}"
```
Comment (adamkdean) [inline test/app/datasources/car-makes-chained-endpoint.json]: What is going on with the spacing here? 😱

File: dadi/lib/controller/index.js (hunk 3 above)
Comment (adamkdean) [reply dadi/lib/controller/index.js]: Ah, ignore me then, couldn't see it in the expanded diff

File: test/app/datasources/car-makes-chained-endpoint.json (hunk 5 above)
Comment (jimlambie) [reply test/app/datasources/car-makes-chained-endpoint.json]: fixed :)

File: dadi/lib/controller/index.js (hunk 4 above)
Comment (jimlambie) [reply dadi/lib/controller/index.js]: done

**Comment by adamkdean** [comment]
🥇

**Pull request merged**